File: core/src/vct/col/ast/Axiom.java
package vct.col.ast;

import java.util.ArrayList;

import vct.util.ClassName;

public class Axiom extends ASTDeclaration {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  private ASTNode rule;
  
  public Axiom(String name, ASTNode exp) {
    super(name);
    rule=exp;
  }

  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
      }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 

  @Override
  public ClassName getDeclName() {
    return new ClassName(name);
  }

  public ASTNode getRule(){
    return rule;
  }
}

File: vercors/src/main/java/vct/col/ast/Axiom.scala
package vct.col.ast

import vct.col.util.VisitorHelper
import vct.util.ClassName

case class Axiom(override val name:String, val rule:ASTNode) extends ASTDeclaration(name) with VisitorHelper {
  override def getDeclName() = new ClassName(name)
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
}
